### ISSUE TITLE
Weavers not found due to case sensitivity, nuget restore

### ISSUE THREAD

**Issue by ColinDabritzViewpoint** [Fody/Fody#331]
Here is our scenario:

- We are building a mobile application via VSTS on a mac
- We are using Fody, and the PropertyChanged weaver
- We found that Fody fails to find the weaver in a lower case package name scenario, outlined below

build output for the Fody MSBuild target:
```
Target FodyTarget:
	Fody: Fody (version 1.29.4.0) Executing
	Fody: ProjectDirectory: '/buildagent/_work/2/s/App/DeveloperModule/'.
	Fody: AssemblyPath: '/buildagent/_work/2/s/App/DeveloperModule/obj/Release/DeveloperModule.dll'
	Fody: Found path to weavers file '/buildagent/_work/2/s/App/DeveloperModule/FodyWeavers.xml'.
	Fody: SolutionDirectory path is '/buildagent/_work/2/s/App/'
	Fody: Finding weavers
	Fody: Adding weaver dlls from '/buildagent/_work/2/s/App/Packages'.
	Fody: Could not find packages dir from nuget config.
	Fody: SolutionDirectoryPath: /buildagent/_work/2/s/App/
	Fody: Adding weaver dlls from '/Users/buildagent/.nuget/packages/fody'.
	Fody: SolutionDirectoryPath: /buildagent/_work/2/s/App/
	Fody: Skipped scanning '/buildagent/_work/2/s/App/Tools' for weavers since it doesn't exist.
	Fody: No Weaver project file found.
: error : Fody: Could not find a weaver named 'PropertyChanged'. 
```

After some troubleshooting, we found the following path was problematic: `/Users/buildagent/.nuget/packages/propertychanged.fody`

This is the standard nuget package cache location for the build agent user.

When we renamed the package, using:
`mv propertychanged.fody PropertyChanged.Fody` the build began working.

Unfortunately, the lower case behavior appears to be base behavior in Nuget 3.5.0.1938, which is what our VSTS Step runs, for example:

```
2017-05-10T18:09:18.6094670Z ##[section]Starting: NuGet restore developer module
2017-05-10T18:09:18.6263120Z ==============================================================================
2017-05-10T18:09:18.6275690Z Task         : NuGet Installer
2017-05-10T18:09:18.6288140Z Description  : Installs or restores missing NuGet packages
2017-05-10T18:09:18.6300480Z Version      : 0.2.31
2017-05-10T18:09:18.6312850Z Author       : Microsoft Corporation
2017-05-10T18:09:18.6325370Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613747)
2017-05-10T18:09:18.6337990Z ==============================================================================
2017-05-10T18:09:19.1140180Z Detected NuGet version 3.5.0.1938 / 3.5.0
```

And we are hesitant to explore odd work-arounds like attempting to rename the folder after the fact on a set of shared build machines.

I would expect that locating the PropertyChanged weaver should not be case sensitive, as the ID is not treated as case sensitive by Nuget.

We use Fody heavily, and appreciate this excellent project. Thanks in advance for any assistance with this issue!

**Comment by SimonCropp** [comment]
can u help me out a bit here. does `propertychanged.Fody` or `PropertyChanged.fody` work?

**Comment by ColinDabritzViewpoint** [comment]
TL:DR: `propertychanged.Fody` ** WORKS **, `PropertyChanged.fody` - ERROR

To clarify the structure, the full default package path is:
`/Users/buildagent/.nuget/package/propertychanged.fody/1.52.1/`

containing:

```
Lib/
PropertyChanged.Fody.dll
PropertyChanged.Fody.pdb
Tools/
propertychanged.fody.1.52.1.nupkg
propertychanged.fody.1.52.1.nupkg.sha512
propertychanged.fody.nuspec
```

The .nuspec file specifies the capitalized form of the ID:
`<id>PropertyChanged.Fody</id>`

and the names internally are unchanged in all tests. Just the name of the folder after `package` and before the version in the path are modified.

Here are the four variations I just tested:

- default nuget restore
   - `/Users/buildagent/.nuget/package/propertychanged.fody`
   - ERROR

- Older Nuget, capitalized package ID
    - `/Users/buildagent/.nuget/package/PropertyChanged.Fody`
    - WORKING

- Test case, capitalizing plugin/weaver name
   - `/Users/buildagent/.nuget/package/PropertyChanged.fody`
   - ERROR (same as described before)

- Test case, capitalizing only 'Fody'
   - `/Users/buildagent/.nuget/package/propertychanged.Fody`
   - WORKING

That is unexpected, I was expecting that not matching the package ID exactly would fail, but this is partial. Different code is perhaps matching the first part and the `.Fody` faux extension?

I believe this path name represents the package ID. The Nuspec specifies the capitalized form. I believe not many people run into nuget and build automation on a mac in this way, and the capitalization started mattering. Different versions of nuget have pulled down a capitalized package folder on the mac before.

As a possibly helpful citation, I believe this is recognized, and the intent is for Nuget package identity to be case insensitive and normalized in the cache for this reason. See: https://github.com/NuGet/Home/issues/2522 for example.

If I can be of further help here, please let me know. We're happy to test a pre-release version as well. I'm hosting a build conference livestream locally tomorrow, but I'll try to find time to assist with this.

**Comment by SimonCropp** [comment]
i have had a look at this and cant see where we r case sensitive. can u do a quick tests and see if it occurs on fody 2.0.8?

**Comment by ColinDabritzViewpoint** [comment]
**Summary: No change in behavior after upgrading.**

We were previously running the latest core Fody 1.x package, 1.29.4, and PropertyChanged,Fody version 1.52.1.

I upgraded to Fody 2.0.8, as well as PropertyChanged 1.53.0.

I am still seeing the same issue:

```
Target FodyTarget:
  Fody: Fody (version 2.0.8.0) Executing
  Fody: ProjectDirectory: '/buildagent/_work/2/s/App/DeveloperModule/'.
  Fody: AssemblyPath: '/buildagent/_work/2/s/App/DeveloperModule/obj/Release/DeveloperModule.dll'
  Fody: Found path to weavers file '/buildagent/_work/2/s/App/DeveloperModule/FodyWeavers.xml'.
  Fody: SolutionDirectory path is '/buildagent/_work/2/s/App/'
  Fody: Finding weavers
  Fody: Adding weaver dlls from '/buildagent/_work/2/s/App/Packages'.
  Fody: Could not find packages dir from nuget config.
  Fody: SolutionDirectoryPath: /buildagent/_work/2/s/App/
  Fody: Adding weaver dlls from '/Users/buildagent/.nuget/packages/fody'.
  Fody: SolutionDirectoryPath: /buildagent/_work/2/s/App/
  Fody: Skipped scanning '/buildagent/_work/2/s/App/Tools' for weavers since it doesn't exist.
  Fody: No Weaver project file found.
: error : Fody: Could not find a weaver named 'PropertyChanged'.
```

As before, `PropertyChanged.fody` and `propertychanged.fody` did NOT build successfully, and `PropertyChanged.Fody` and `propertychanged.Fody` did build successfully.

Maybe a dependency being delegated to is unexpectedly case sensitive on Mac?

How else can I help?

**Comment by SimonCropp** [comment]
@ColinDabritzViewpoint i would appreciate if u did a code review of the weaver detection code in fory

**Comment by ColinDabritzViewpoint** [comment]
I think we found it! This is a change in behavior on Mono, and I suspect the same is probably true of dot net core as well. Our environment is executing on mono on mac.

The specific line is this:
https://github.com/Fody/Fody/blob/36897cce6858e0e90830c21f4041f4618dadcc86/Fody/AddinFinder/AddinSearchDirectories.cs#L42
`foreach (var packageDirectory in Directory.EnumerateDirectories(nuGetPackageRoot, "*.Fody"))`

And there is an open Xamarin bug on this here: https://bugzilla.xamarin.com/show_bug.cgi?id=28875

I can see a couple approaches to addressing this issue, such as searching for `"*.?ody"`, or as suggested in the bug filing, dropping the pattern and filtering the results manually in a "where" clause.

We're happy to try a beta package to confirm this fix.

**Comment by ColinDabritzViewpoint** [comment]
I had a chance to look into this further, and I put together a pull request to make the search case insensitive across platforms. Let me know if any further changes would be useful, and feel free to adjust or format as desired.

Since this change technically increases the matching scope slightly and may affect existing projects, but generally should be backward compatible, I recommend a minor version revision to the package such as 2.1.0.

I appreciate your help and responsiveness around this issue, despite not having a mac environment available!

**Comment by ColinDabritzViewpoint** [comment]
I updated to 2.0.10 on our project and built with lowercase packages present. Thanks for your help with this issue, and thanks for making Fody a reality!

**Comment by BjornerJF** [comment]
This is still an issue, Visual Studio 2017 on Windows 10 build 1709.

**Comment by SimonCropp** [comment]
@BjornerJF upload a repro solution to a github repository

**Comment by ColinDabritzViewpoint** [comment]
Windows should be case insensitive, it was never an issue on windows in my scenario. It only became an issue using it on Xamarin Studio (now Visual Studio for Mac) on OSX. Perhaps the problem is related, or just a similar symptom?

**Comment by P250** [comment]
Yeah I'h having this problem too in Visual Studio 2017, how STUPID.

**Comment by SimonCropp** [comment]
@P250 can u share a repro

**Comment by RezaAshena** [comment]
I just add as existing item from another project.and fixed my bug
